checkmate/url/canonicalize.py

```diff
@@ -22,6 +22,25 @@ def canonicalize(cls, url):
         :param url: URL to normalise
         :return: A normalised version of the URL
         """
+        parts = cls.canonical_split(url)
+        clean_url = urlunparse(parts)
+
+        # Get around the fact that URL parse strips off the '?' if the query
+        # string is empty
+        if url.endswith("?") and not clean_url.endswith("?"):
+            clean_url += "?"
+
+        return clean_url
+
+    @classmethod
+    def canonical_split(cls, url):
```

Comment:
This function is bonkers fast. To the point I'm not sure if I'm timing it right, but I can do this around 5 million times _per millisecond_, so I don't think we have to worry about calling this as much as we like.

I was afraid this was slow, but I think the slow part is the SHA hash.